include/vctr/Expressions/Generic/TransformedByDynamicCast.h

```diff
@@ -0,0 +1,63 @@
+/*
+  ==============================================================================
+    DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+
+    Copyright 2025 by sonible GmbH.
+
+    This file is part of VCTR - Versatile Container Templates Reconceptualized.
+
+    VCTR is free software: you can redistribute it and/or modify
+    it under the terms of the GNU Lesser General Public License version 3
+    only, as published by the Free Software Foundation.
+
+    VCTR is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU Lesser General Public License version 3 for more details.
+
+    You should have received a copy of the GNU Lesser General Public License
+    version 3 along with VCTR.  If not, see <https://www.gnu.org/licenses/>.
+  ==============================================================================
+*/
+
+namespace vctr::expressions
+{
+
+template <class T>
+concept isRefOrPtr = is::lvalueReference<T> || is::pointer<T>;
+
+template <size_t extent, class SrcType, isRefOrPtr DstValueType, is::constant ExpectCastToSucceed>
+class TransformedByDynamicCast : public ExpressionTemplateBase
```

Comment:
While you are implementing this. Could you implement a `transformedBy `function that is also able to take an index of the current element?